src/prepared_discover/prepared_discover_walk.c

```diff
@@ -0,0 +1,469 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __prepared_discover_btree_has_prepare --
+ *     Check the metadata entry for a btree to see whether it included prepared updates.
+ */
+static int
+__prepared_discover_btree_has_prepare(WT_SESSION_IMPL *session, const char *config, bool *has_prepp)
+{
+    WT_CONFIG ckptconf;
+    WT_CONFIG_ITEM cval, key, value;
+    WT_DECL_RET;
+
+    *has_prepp = false;
+
+    /* This configuration parsing is copied out of the rollback to stable implementation */
+    WT_RET(__wt_config_getones(session, config, "checkpoint", &cval));
+    __wt_config_subinit(session, &ckptconf, &cval);
+    for (; __wt_config_next(&ckptconf, &key, &cval) == 0;) {
+        ret = __wt_config_subgets(session, &cval, "prepare", &value);
+        if (ret == 0) {
+            if (value.val)
+                *has_prepp = true;
+        }
+        WT_RET_NOTFOUND_OK(ret);
+    }
+    return (0);
+}
+
+/*
+ * __prepared_discover_process_ondisk_kv --
+ *     Found an on-disk prepared value, process it into a transaction structure.
+ */
+static int
+__prepared_discover_process_ondisk_kv(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip,
+  uint64_t recno, WT_ITEM *row_key, WT_CELL_UNPACK_KV *vpack)
+{
+    WT_CURSOR *hs_cursor;
+    WT_DECL_RET;
+    WT_ITEM *key;
+    WT_PAGE *page;
+    WT_TIME_WINDOW *hs_tw, *tw;
+    uint8_t *memp;
+
+    page = ref->page;
+    if (rip != NULL) {
+        if (row_key != NULL)
+            key = row_key;
+        else {
+            /* Unpack a row key. */
+            WT_ERR(__wt_scr_alloc(session, 0, &key));
+            WT_ERR(__wt_row_leaf_key(session, page, rip, key, false));
+        }
+    } else {
+        /* Manufacture a column key. */
+        WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
+        memp = key->mem;
+        WT_ERR(__wt_vpack_uint(&memp, 0, recno));
+        key->size = WT_PTRDIFF(memp, key->data);
+    }
+
+    /* Retrieve the time window from the unpacked value cell. */
+    __wt_cell_get_tw(vpack, &tw);
+
+    /* Add an entry for this key to the transaction structure */
+    if (rip != NULL)
+        WT_ERR(__wti_prepared_discover_add_artifact_ondisk_row(session, tw->start_ts, tw, key));
+    else
+        WT_ASSERT_ALWAYS(
+          session, false, "Column store prepared transaction discovery not supported");
+
+    /*
+     * Check to see if the transaction made multiple changes to the same key - they will be in the
+     * history store.
+     */
+    WT_ERR(__wt_curhs_open(session, NULL, &hs_cursor));
+    /*
+     * Rollback-to-stable operates exclusively (i.e., it is the only active operation in the system)
```

Comment:
I don't understand why we need to do this here. We never insert prepared updates to the history store. If there are multiple prepared updates on the same key, we insert the first key to the data store and ignore the rest. Prepared updates cannot use different timestamps so they have the same visibility. The new update just mask the older updates on the same key.